engine/api/src/main/java/io/deephaven/engine/table/Table.java

```diff
@@ -89,6 +91,39 @@ public interface Table extends
     @ConcurrentMethod
     boolean isRefreshing();
 
+    @ConcurrentMethod
+    UpdateContext getUpdateContext();
+
+    /**
+     * Ensure that the table is compatible with the currently active update context.
+     *
+     * @param others Other source tables to check for consistency
+     */
+    @FinalDefault
+    default void checkUpdateContextConsistency(final Table... others) {
+        UpdateContext foundUpdateContext = null;
+        if (isRefreshing()) {
+            foundUpdateContext = getUpdateContext();
+        }
+
+        for (final Table other : others) {
+            if (other == null || !other.isRefreshing()) {
+                continue;
+            }
+            if (foundUpdateContext == null) {
+                foundUpdateContext = other.getUpdateContext();
+            } else if (foundUpdateContext != other.getUpdateContext()) {
+                throw new IllegalStateException("Tables from different update contexts cannot be mixed. Found: "
+                        + foundUpdateContext + " and " + other.getUpdateContext());
+            }
+        }
+
+        if (foundUpdateContext != null && foundUpdateContext != UpdateContext.get()) {
```

Comment:
Why do we care about this? We care about using the logical clock from the tables' UGP, and/or the locks from their UGP. I think only new roots should be finding their UGP from the context, derived tables should be inheriting.